chore: reduce scheduler complexity [DET-4082]

shiyuann opened the pull request:
## Description

We have built gang scheduling, GPU/CPU task scheduling, fair share, and priority scheduling into the scheduler. As we added more features to the system, we introduced much complexity as well. This complexity mainly is on the following perspectives:

1. The scheduler serves many orthogonal functionalities, including registering/unregistering proxy, scheduling, failure handling for a gang task, container starting and terminating, task termination. Some of them are useful only for trials.
2. The protocol between the resource providers and the task actors is very complex. For example, gracefully canceling a trial needs to send at least 4 messages from and to the resource providers. Also, the task actor typically maintains a state machine. In order to make the state machines in the task actor and the scheduler consistent, we need to carefully avoid race conditions on state synchronization.
3. We currently have two resource providers and need to maintain consistency between them in front of the task actors, which is barely possible because the protocol is so complex and in some cases unclear what the inputs and outputs are.

The negative impacts of this complexity are:

1. The cognitive cost of understanding the code is huge. In some cases, we know the code work but we don't know why unless we spend a huge amount of time.
2. The maintenance cost of making two resource providers consistent is big.
3. The scheduler becomes soft blockers for new features or fixing bugs. We need delicate workarounds in order to add new features or fix bugs. Some known bugs are as follows:

   - [DET-3792](https://determinedai.atlassian.net/browse/DET-3792): in order to ensure cleanup of containers always happen, we always clean up when a scheduler task is terminated just because we were maintaining two states of a single task both on the task and the scheduler and we are not sure if they are synced.
   - [DET-3802](https://determinedai.atlassian.net/browse/DET-3802): the scheduler get information from both the agents and the tasks so sometimes the scheduler doesn't know which actor message represents the up-to-date state. Since we maintained a state machine inside the scheduler, we can't know if the state can be transitioned properly by checking the transition table. We introduced a hack to allow the state to be transitioned from `Terminated` to `Terminated` to fix a bug, which might fundamentally break the assumption we made for the state machines.
   - [DET-3643](https://determinedai.atlassian.net/browse/DET-3643): It's hard to add time-aware scheduling since the protocol is already complicated.
   - It's hard to add across-resource-pool logic. We will introduce the concept of resource pools, each of them contains one scheduler. We will also introduce an across-resource pool scheduling level to support the use case of one task using multiple resource pools. Adding fair-share and d-train failure handling would be very hard.

4. Poor testability and debuggability.

   - It's hard to add fine-grained automatic tests.
   - Too many unknown unknowns. In most cases, we identify the issues when users actually use it. And it's hard to tell the cause of the issues by simply adding logging.
      - We have observed some issues including the scheduler staring pending tasks, not handling task exiting well, asking the provisioned for more instances while having enough slots, stale tasks after tasks fail, deadlock problems: [DET-3642](https://determinedai.atlassian.net/browse/DET-3642), [DET-3545](https://determinedai.atlassian.net/browse/DET-3545), [DET-3523](https://determinedai.atlassian.net/browse/DET-3523), [DET-1196](https://determinedai.atlassian.net/browse/DET-1196), [DET-1107](https://determinedai.atlassian.net/browse/DET-1107)

The new design of the scheduler in this PR follows such principles:

1. It should only care what resources it has and which agent they belong to, which tasks occupy which resources, and which task group each task belongs to.
2. It shouldn't care containers starting and terminating, task state, and proxy.

This PR aims at removing the container state and task state in the resource providers.

## Test Plan

1. Use the test suite. https://app.circleci.com/pipelines/github/determined-ai/determined?branch=scheduler-debt
2. Manually test canceling/pausing/killing an experiment when it's active and pending.
3. Manually test killing a command/Tensorboard/Notebook.
4. Manually run multiple tasks and test if the fair-share and priority scheduler work.

## Commentary (optional)

N/A

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

shiyuann commented:
Please ignore trial.go for now since I am still trying to fix trial aborting for it.

shiyuann commented:
@aaron276h fix a couple of bugs on handling aborted resource requests and zero-slot resources. Now, it's ready for reviewing.

aaron276h reviewed (commented):
@shiyuann i haven't gone through the entire PR here but there are enough comments to get it back to you. I like the overall approach but there are some implementation details that need fixing up. 

I also don't really like the rename of Task -> request, but not dead set on not renaming it.

shiyuann commented on master/internal/kubernetes/pod.go:
```diff
@@ -250,11 +254,28 @@ func (p *pod) receivePodStatusUpdate(ctx *actor.Context, msg podStatusUpdate) er
+			ContainerInfo: types.ContainerJSON{
+				ContainerJSONBase: &types.ContainerJSONBase{
+					HostConfig: &docker.HostConfig{
+						NetworkMode: "host",
```
removed

shiyuann commented on master/internal/kubernetes/pod.go:
```diff
@@ -250,11 +254,28 @@ func (p *pod) receivePodStatusUpdate(ctx *actor.Context, msg podStatusUpdate) er
+		// This is a hack to construct a Docker container info struct for pod since
+		// the Docker container info struct is used in the protocol between the resource
+		// provider and the task handler.
+		transPort := func(ports []int) nat.PortSet {
```
removed.

shiyuann commented on master/internal/kubernetes/request_queue.go:
```diff
@@ -137,6 +137,7 @@ func (r *requestQueue) Receive(ctx *actor.Context) error {
 				return errors.Errorf("%s actor already exists", newWorker.Address())
 			}
 		}
+	case actor.PostStop:
```
Every actor receives an `actor.PostStop`. When I was debugging k8s and the master errored out, this actor receives `actor.PostStop` message. Also I think actor messages should be grouped together by its sender

shiyuann replied:
Added a comment to explain it shouldn't execute to there.

shiyuann commented on master/internal/kubernetes/request_workers.go:
```diff
@@ -16,6 +16,7 @@ func (r *requestProcessingWorker) Receive(ctx *actor.Context) error {
 	switch msg := ctx.Message().(type) {
 	case actor.PreStart:
 		ctx.Tell(ctx.Self().Parent(), workerAvailable{})
+	case actor.PostStop:
```
same.

shiyuann replied:
Added a comment to explain it shouldn't execute to there.

shiyuann commented on master/internal/trial.go:
```diff
@@ -281,6 +287,17 @@ func (t *trial) Receive(ctx *actor.Context) error {
 
+	case trialAborted:
+		// This is a hack to handle trial being aborted. Previously, the scheduler
+		// sends TaskTerminated and TaskAborted message to notify trial that the
```
fixed

shiyuann commented on master/internal/trial.go:
```diff
@@ -281,6 +287,17 @@ func (t *trial) Receive(ctx *actor.Context) error {
 	case sproto.ContainerLog:
 		t.processLog(ctx, msg)
 
+	case trialAborted:
```
Unfortunately, the trial actor sends a message to itself but it does something. It has the logic to handle trial closing, adding a task to the scheduler, and gracefully termination right below the switch statement in `Receive`.

aaron276h replied:
Ok i think we should mention this in the comment below.

shiyuann commented on master/internal/scheduler/kubernetes_resource_provider.go:
```diff
@@ -282,169 +246,25 @@ func (k *kubernetesResourceProvider) getOrCreateGroup(
+		req := it.value()
+		group := k.groups[req.Group]
+		assigned := k.reqList.GetAssignments(req.Handler)
+		if assigned == nil || len(assigned.Assignments) == 0 {
```
done.

shiyuann commented on master/internal/scheduler/agent.go:
```diff
@@ -11,19 +11,22 @@ import (
+
+	// This is a hack for letting the provisioner knows which agents are being used
+	// since we don't model CPU as a resource type.
+	zeroSlotContainers map[*cproto.ID]bool
```
Talked offline. we will keep this as is in the PR.

shiyuann commented on master/internal/scheduler/request.go:
```diff
@@ -0,0 +1,41 @@
+	AssignRequest struct {
+		ID                  RequestID
+		Name                string
+		Group               *actor.Ref
```
This shouldn't use `group` since `group` is an internal state in the scheduler.

shiyuann commented on master/internal/scheduler/default_resource_provider.go:
```diff
@@ -302,256 +264,46 @@ func (d *DefaultRP) receiveAgentMsg(ctx *actor.Context) error {
-		addresses := toAddresses(
-			msg.ContainerStarted.ProxyAddress, msg.ContainerStarted.ContainerInfo)
-		d.receiveContainerStartedOnAgent(ctx, cid, addresses)
+	case sproto.ContainerStateChanged:
```
removed.

shiyuann commented on master/internal/trial.go:
```diff
@@ -1056,12 +1078,22 @@ func (t *trial) trialClosing() bool {
 func (t *trial) terminate(ctx *actor.Context, kill bool) {
-	if kill {
+	switch {
+	case len(t.assignments) == 0:
```
Good question. We used not to handle that. So I added a commit to ignore the assigned message if it finds the task doesn't exist or the task ID is not matched.

shiyuann commented on master/internal/scheduler/request_list.go:
```diff
@@ -0,0 +1,96 @@
+	return req, ok
+}
+
+func (l *assignRequestList) Add(req *AssignRequest) bool {
```
making it exported for now since we will soon move things around

shiyuann commented on master/internal/command/command.go:
```diff
@@ -282,8 +285,14 @@ func (c *command) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
-	ctx.Ask(c.rps, scheduler.TerminateTask{TaskID: c.taskID, Forcible: true}).Get()
-	if msg, ok := ctx.Message().(scheduler.TerminateRequest); ok {
+	ctx.Log().Info("terminating")
+	ctx.Tell(c.rps, scheduler.ResourceReleased{Handler: ctx.Self()})
```
Fixed that by ignoring the allocated messages if the allocated messages use different task IDs as the trial and the command currently use

shiyuann replied:
talked offline. Rather than sending `ResourcesReleased` right after killing the containers, we do that when receving the container state changed messages.

shiyuann commented on master/internal/command/command.go:
```diff
@@ -100,19 +101,22 @@ func (c *command) Receive(ctx *actor.Context) error {
 		})
 		ctx.Tell(c.eventStream, event{Snapshot: newSummary(c), ScheduledEvent: &c.taskID})
 
+	case actor.PostStop:
```
Added a comment to explain it shouldn't execute to there.

shiyuann commented on master/internal/checkpoint_gc.go:
```diff
@@ -27,15 +27,15 @@ type checkpointGCTask struct {
 		})
 
-	case scheduler.TaskAssigned:
+	case scheduler.ResourceAssigned:
```
Talked offline. We use `ResourceAllocated`

shiyuann commented on master/internal/checkpoint_gc.go:
```diff
@@ -27,15 +27,15 @@ type checkpointGCTask struct {
 				SingleAgent: true,
 			},
-			TaskHandler: ctx.Self(),
+			Handler: ctx.Self(),
```
Talked offline. Settled to use `TaskActor` rather than `Handler`

aaron276h commented on master/internal/scheduler/default_resource_provider.go:
```diff
@@ -80,92 +58,73 @@ func NewDefaultRP(
+	// the old container exiting and the new container run slower than normal.
+	// The task handler should kill the containers to physically release the resources
+	// after a timeout.
+	d.taskList.RemoveTaskByHandler(handler)
```
question: why do we have this here and not just only have in `resourcesReleased()`?

shiyuann replied:
This is not an accurate comment anymore. Also removed this line.

aaron276h commented on master/internal/scheduler/default_resource_provider.go:
```diff
@@ -80,92 +58,73 @@ func NewDefaultRP(
+	// In practice, the request handler might gracefully wait for the containers
+	// to exit themselves, which might take very long time or might not happen at all.
+	// In the mean time, resources are re-allocated to other requests, which might let
+	// the old container exiting and the new container run slower than normal.
```
not really clear about what it means that new containers will run slower, does it mean they will get scheduled on the same slot as a task that not yet exited?

shiyuann replied:
This is not accurate anymore because I fixed when containers are killed a slot might be allocated to two containers.

shiyuann replied:
> not really clear about what it means that new containers will run slower, does it mean they will get scheduled on the same slot as a task that not yet exited?

yes. I removed the comment and the call to `RemoveTaskByHandler`

aaron276h commented on master/internal/scheduler/filterable_view.go:
```diff
@@ -25,14 +25,16 @@ func newProvisionerView(provisionerSlotsPerInstance int) *FilterableView {
 
 		switch {
-		case task.state != taskPending:
+		case assigned != nil && len(assigned.Allocations) > 0:
```
non-bcloking: maybe add a helper function to task_list for `isAssigned()`

shiyuann replied:
I found this is the only place that needs to determine if a task is allocated. I will rewrite this part anyways when I refactor the protocol between the scheduler and the provisioner. So won't bother doing that now.

aaron276h reviewed (commented):
@shiyuann I haven't had yet a chance to go through `trial.go` yet so I am going to keep this assigned to myself (planning to do it Monday morning), but wanted to post the comments I have so far. 

This is looking good, nice work!

shiyuann commented on master/internal/checkpoint_gc.go:
```diff
@@ -46,18 +47,22 @@ func (t *checkpointGCTask) Receive(ctx *actor.Context) error {
-			})
+		for _, a := range msg.Allocations {
+			taskSpec := tasks.TaskSpec{}
+			if t.defaultTaskSpec != nil {
```
I should have explained it in this PR. So I removed the definition of the `Container` interface and overriding some values in `containerAssignment.StartContainer`, which requires to remove `harnessPath`, `ClusterID`, and `taskDefaults` out of the scheduler.

shiyuann replied:
Currently, we always have `defaultTaskSpec` to be not `nil`. I'll change to be `taskSpec` for all task actors.

shiyuann commented on master/internal/command/command.go:
```diff
@@ -281,11 +268,74 @@ func (c *command) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
+	if c.allocation == nil {
+		c.exit(ctx, "command is terminated without being scheduled")
+	} else {
+		ctx.Log().Info("forcible terminating")
```
I thought here we should use "command" since there is no more "terminating" concept in the scheduler.

shiyuann replied:
changed to `task forcibly terminating`

shiyuann commented on master/internal/command/command.go:
```diff
@@ -281,11 +268,74 @@ func (c *command) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
 	}
+
+	if c.allocation == nil {
+		c.exit(ctx, "command is terminated without being scheduled")
```
I thought here we should use "command" since there is no more "terminating" concept for tasks in the scheduler.

shiyuann replied:
Okay. I changed that to be `task is aborted without being scheduled`

shiyuann commented on master/internal/kubernetes/spec.go:
```diff
@@ -346,16 +346,16 @@ func (p *pod) createPodSpecForGC(ctx *actor.Context) error {
+			"exp-%d-trial-%d-%s",
 			t.StartContainer.InitialWorkload.ExperimentID,
-			t.StartContainer.InitialWorkload.TrialID, rank,
+			t.StartContainer.InitialWorkload.TrialID,
```
I can print the rank in `trial.go` since I removed the rank concept from the resource providers. `rank` is more of a concept that is being used outside the scheduler. The scheduler doesn't and shouldn't care which assignment is the leader.

shiyuann replied:
> The scheduler is actually the one which implicitly assigns rank to assignments when it creates them

I changed that to be the order of containers connected. why is it necessary to let the scheduler implicitly handles that?

shiyuann replied:
changed that to be the previous state

shiyuann commented on master/internal/agent/slots.go:
```diff
@@ -81,7 +81,15 @@ func (s *slots) summarize(ctx *actor.Context) SlotsSummary {
 func (s *slots) sendToSlots(ctx *actor.Context, c container.Container, msg actor.Message) {
-	for _, d := range c.Devices {
-		ctx.Tell(ctx.Child(d.ID), msg)
+	if len(c.Devices) == 0 && c.State == container.Terminated {
```
The reason is that the scheduler needs a message to know which resources are released. For GPU devices, it receives messages `FreeDevice` from the slot actors. We model CPU now as a special case and we still need a way to tell the scheduler when a container is terminated. So I reuse the `FreeDevice` message here and let the `slots` actor notify the scheduler of the container termination.

shiyuann commented on master/internal/kubernetes/pods.go:
```diff
@@ -238,10 +235,10 @@ func (p *pods) startResourceRequestQueue(ctx *actor.Context) {
 
 func (p *pods) receiveStartPod(ctx *actor.Context, msg sproto.StartPod) error {
```
Changed the names to be `StartTaskContainer` for the default resource provider and `StartTaskPod` for the k8s resource provider

shiyuann commented:
@aaron276h I added a commit for rewriting the unit tests for fair share and priority after I broke the tests when I removed removing the task right after the scheduler ask the task actor to release resources.

aaron276h commented on master/internal/trial.go:
```diff
@@ -899,9 +935,11 @@ func (t *trial) processContainerTerminated(
 		t.terminate(ctx, true)
 	}
+
+	t.terminated(ctx)
```
blocking: I don't think we are handling distributed scenarios correctly here because right now we are are terminating the trial as soon as one of the containers exits out.

shiyuann replied:
fixed it.

shiyuann commented on master/internal/trial.go:
```diff
@@ -780,20 +796,24 @@ func (t *trial) allReady(ctx *actor.Context) bool {
 // pushRendezvous gathers up the external addresses for the exposed ports and sends them to all the
 // containers in the trial.
 func (t *trial) pushRendezvous(ctx *actor.Context) error {
+	ctx.Log().Info("pushing rendezvous information")
```
I think it's useful when debugging if a trial is an unscheduled/scheduled state since we don't reflect that in its state.

shiyuann commented on master/internal/trial.go:
```diff
@@ -921,7 +959,75 @@ func (t *trial) processLog(ctx *actor.Context, msg sproto.ContainerLog) {
+			}
+		}
+	default:
+		ctx.Log().Info("gracefully terminating trial")
```
Added a case "!t.pendingGracefulTermination" in the switch statement.

shiyuann commented on master/internal/trial.go:
```diff
@@ -921,7 +959,75 @@ func (t *trial) processLog(ctx *actor.Context, msg sproto.ContainerLog) {
+		t.terminated(ctx)
+		ctx.Tell(ctx.Self(), trialAborted{})
+	case kill:
+		ctx.Log().Info("forcibly terminating trial")
```
Exactly and this PR doesn't aim at fixing that due to the complexity of `trial.go`. However, reducing the scheduler complexity helps us fix that later in the future. :)

shiyuann commented on master/internal/trial.go:
```diff
@@ -489,7 +495,18 @@ func (t *trial) processID(ctx *actor.Context, id int) {
-func (t *trial) processAssigned(ctx *actor.Context, msg scheduler.TaskAssigned) error {
+func (t *trial) processAllocated(ctx *actor.Context, msg scheduler.ResourcesAllocated) error {
+	// Ignore this message if the trial is terminated or the message is from the last run of the trial.
+	if t.task == nil {
```
we don't. We actually handling sending `ResourceReleased` in `terminated`. :) This is to handle the race condition we were discussing. A trial is terminated doesn't mean it exits/closes. We handle restarting logic in `terminated`.

shiyuann commented on master/pkg/device/device.go:
```diff
@@ -23,6 +25,8 @@ func (t Type) Proto() devicev1.Type {
 		return devicev1.Type_TYPE_CPU
 	case GPU:
 		return devicev1.Type_TYPE_GPU
+	case Unspecified:
```
fixed it to `ZeroSlot` as discussed

shiyuann commented on master/internal/agent/slots.go:
```diff
@@ -81,7 +81,15 @@ func (s *slots) summarize(ctx *actor.Context) SlotsSummary {
+	if len(c.Devices) == 0 && c.State == container.Terminated {
+		ctx.Tell(s.cluster, sproto.FreeDevice{
+			DeviceID: sproto.DeviceID{
+				Agent: ctx.Self().Parent(), Device: device.Device{Type: device.Unspecified}},
```
fixed it to `ZeroSlot`

aaron276h commented on master/internal/trial.go:
```diff
@@ -871,14 +886,37 @@ func (t *trial) pushRendezvous(ctx *actor.Context) error {
+	_, ok := t.containers[msg.Container.ID]
+	delete(t.containers, msg.Container.ID)
+	delete(t.containerAddresses, msg.Container.ID)
+	delete(t.containerRank, msg.Container.ID)
```
blocking: we are deleting `t.containerRank` here and then using it L931

aaron276h commented on master/internal/trial.go:
```diff
@@ -871,14 +886,37 @@ func (t *trial) pushRendezvous(ctx *actor.Context) error {
+
+	t.containers[msg.Container.ID] = msg.Container
+	t.containerAddresses[msg.Container.ID] = msg.ContainerStarted.Addresses
+	t.containerRank[msg.Container.ID] = len(t.containerRank)
```
blocking: we shouldn't be assigning rank based on the order in which containers connect. The chief container needs to match up with the container that is running process 0.

shiyuann replied:
I changed it to assign rank based on the index of `Allocation` and add the rank back to the name of pods.

shiyuann commented on master/internal/command/command.go:
```diff
@@ -281,11 +268,71 @@ func (c *command) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
+			return nil
+		}
+
+		check.Panic(check.Equal(len(msg.Allocations), 1,
```
I think here is a good place to use panic since it should be definitely our fault if more than one allocations are sent to the command and we can catch them early by using panic here. Reference: https://gobyexample.com/panic#:~:text=Go%20by%20Example%3A%20Panic,-A%20panic%20typically&text=This%20is%20the%20only%20program%20on%20the%20site%20designed%20to%20panic.&text=A%20common%20use%20of%20panic,when%20creating%20a%20new%20file.

aaron276h replied:
In our system we can return an error or panic here, unfortunately we do not have consistency although in the majority of cases we do return errors instead of panics which is why I suggested we make this change.

shiyuann replied:
talked offline. Generally, we should be consistent, which we aren't. According to the convention of using errors and panics based on whether they are expected or caused by users' wrong input, we should use panic here.

shiyuann commented on master/internal/command/command.go:
```diff
@@ -281,11 +268,71 @@ func (c *command) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
+	switch msg := ctx.Message().(type) {
+	case scheduler.ResourcesAllocated:
+		// Ignore this message if the command has exited.
+		if c.task == nil || msg.ID != c.task.ID {
```
It shouldn't happen at all but since the message contains an ID I just checked it

aaron276h commented on master/internal/kubernetes/pods.go:
```diff
@@ -106,8 +103,8 @@ func (p *pods) Receive(ctx *actor.Context) error {
 
-	case sproto.StartPod:
-		if err := p.receiveStartPod(ctx, msg); err != nil {
+	case sproto.StartTaskPod:
```
blocking: can we make `StartTaskPod` and and `KillTaskContainer` messages naming consistent. I would suggest that both RP define their own message type in`sproto` for `KillTaskContainer`. This way task handler call `KillContainer` and then the RP abstracts away that whether it is a pod `KillTaskPod` or a container `KillTaskContainer`. I do realize that both messages have the same exact content which makes it a bit repetitive, but the consistency with how the Start message are names and implemented will improve code readability.

aaron276h commented on master/pkg/device/device.go:
```diff
@@ -14,6 +14,8 @@ const (
 	CPU Type = "cpu"
 	// GPU represents a GPU device.
 	GPU Type = "gpu"
+	// ZeroSlot represents a unspecified device.
```
non-blocking: `ZeroSlot represents cpu devices on agents where only GPUs are modeled.`

aaron276h reviewed (approved):
Awesome work @shiyuann!!! I really like how this refactor has turned out. Left a couple comments but otherwise it is time to 🚀 🌠
